### ISSUE TITLE
Enhancement: Add a [consider-using-join]-checker 

### ISSUE THREAD

**Issue by pheanex** [PyCQA/pylint#1952]
**I suggest adding a checker for unidiomatic concatenation of strings with a for-loop**

Example:
```python
# bad
result = ''
for letter in ['a', 'b', 'c']:
    result += letter

# good
result = ''.join(['a', 'b', 'c'])
```
The idiomatic way would be to use a ''.join()"-statement.
See i.e. [docs.python-guide.org](http://docs.python-guide.org/en/latest/writing/style/#create-a-string-from-a-list) or [blogpost by jeffknupp](https://jeffknupp.com/blog/2012/10/04/writing-idiomatic-python/). (Thanks @jeffknupp)

For functional test-cases i came up with the following (feel free to suggest additions):

```python
# Variations of 'result'
result = ''  # [consider-using-join]
for number in ['1', '2', '3']:
    result += number

result = 'header'  # [consider-using-join]
for number in ['1', '2', '3']:
    result += number

result = another_result = ''  # [consider-using-join]
for number in ['1', '2', '3']:
    result += number

result = 0  # result is not a string
for number in ['1', '2', '3']:
    result += number

reslut = ''  # wrong name / initial variable missing
for number in ['1', '2', '3']:
    result += [number]

string_variable = ''
result = string_variable  # type of 'result' not obviously a string
for number in ['1', '2', '3']:
    result += number

result = ''
another_result = ''  # result defined too early
for number in ['1', '2', '3']:
    result += [number]

for number in ['1', '2', '3']:  # 'result'-definition missing
    result += number


# Variations of 'number'
result = ''  # no concatenation (iterator-name differs)
for name in ['1', '2', '3']:
    result += number

result = ''  # no concatenation (iterator-name differs)
for _ in ['1', '2', '3']:
    result += number
  # 'exprlist' is not a single name
for index, number in ['1', '2', '3']:
    result += number


# Variations of 'iterable'
result = ''  # [consider-using-join]
for number in []:
    result += number

result = ''  # [consider-using-join]
for number in "a text":
    result += number

result = ''  # [consider-using-join]
for number in [1, 2, 3]:
    result += number

a_list = [1, 2, 3]
result = ''  # [consider-using-join]
for number in a_list:
    result += number

result = ''  # [consider-using-join]
for number in ['1', '2', '3']:
    result += number

result = ''  # [consider-using-join]
for number in undefined_iterable:
    result += number


# Variations of loop-body
result = ''  # addition is not the only part of the body
for number in ['1', '2', '3']:
    print(number)
    result += number

result = ''  # addition is not the only part of the body
for number in ['1', '2', '3']:
    result += number
    print(number)

result = ''  # augmented addition is not a simple one
for number in ['1', '2', '3']:
    result += '4' + number

result = ''  # assignment is not augmented
for number in ['1', '2', '3']:
    result = number

result = ''  # augmented assignment is not an addition
for number in ['1', '2', '3']:
    result -= number

result = ''  # addition is not the 'number'-iterable
for number in ['1', '2', '3']:
    result += another_number

result = ''  # [consider-using-join]
for number in ['1', '2', '3']: result += number
```
For the implementation of the checker i was thinking of looking for string-variable assignments followed by a specific for loop-construct with a particular body.

How do you feel about this enhancement?
Thoughs, comments, objections?

**Comment by PCManticore** [comment]
While this pattern of joining strings is indeed not idiomatic, I don't actually think it is actually used that much in the wild. Unless this is a gotcha that trips beginners or folks in general, I wouldn't see a lot of benefit for adding it straight into pylint. If you have some examples though where this pattern is prevalent, I would be happy to reconsider this issue.

**Comment by pheanex** [comment]
Thank you for the feedback.

Of course I do not have any numbers or estimates on how often this actually occurs in the whole python eco-system.
I personally have seen it often enough to suggest and implement this checker, as I think it's a help to beginners and people who want to improve their code (and rely on linters to get better).

Another indicator that this might be of value is the fact that this idiom is mentioned in almost every article, blog or book about "how to write idiomtic python" I came across until today.
Also I have not found any other linters yet, that tackle this issue. (If anyone knows of any, please comment)

If you nevertheless do not want to add so many idiomatic-checkers to pylint directly (I was planning on proposing/adding quite a few more in the near future :smile:), I'd suggest a compromise to add this one (and others) as an "idiomatic"-extension. How do you/others feel about this idea?

**Comment by PCManticore** [comment]
Alright @pheanex this sounds reasonable after all, thanks for the arguments. Regarding the check itself, I'd implement it from the augmented assignment upwards, rather than the other way around, since this will be easier to implement.

**Comment by pheanex** [comment]
Ok, would you rather have me put this (and other idiomatic-checkers) into a sperate extension or pylint (refactoring) directly?
Thank you for the implementation tip.

**Comment by PCManticore** [comment]
Let's start with core first and see where that takes us. Unfortunately the extensions seems right now like a place where checkers go to die, since it doesn't seem that a lot of folks are familiar with them (that is a point to be made wrt documentation).
